Remove source from modelDB

hbcarlos opened the pull request:
## References
* #11434
* #11602

## Code changes
Removes `value` from ModelDB to use its counterpart `source` from the shared model:
https://github.com/jupyterlab/jupyterlab/blob/014091f21d00bd1c940105744d945a61a7b23e6a/packages/codeeditor/src/editor.ts#L244-L246
https://github.com/jupyterlab/jupyterlab/blob/014091f21d00bd1c940105744d945a61a7b23e6a/packages/shared-models/src/ymodels.ts#L98

## User-facing changes
N/A

## Backwards-incompatible changes
* Regarding the `IModel`, I removed the property `value`. Now we can interact with the content with:
    * `model.value.changed` -> `model.sourceChanged`
    * `model.value.text` -> `model.source`
    * `model.value.remove` -> `model.source = ""`
    * `model.value.insert` -> `model.updateSource`

hbcarlos commented:
I'm not sure if the best solution is keeping the `value` property in the `IModel` because until now, the `value` was an `IObservableString` which allows add/remove/update/listen. Now all these operations are in the `sharedModel` and the `value` is just a get/set to replace the content of the source in the shared model.

Maybe the best solution is just removing the property `value` and access to `sharedModel` instead:
* `model.value.changed` -> `model.sharedModel.changed`
* `model.value.text` -> `model.sharedmodel.getSource`
* `model.value.text` -> `model.sharedmodel.setSource`
* `model.value.remove` -> `model.sharedModel.updateSource`
* `model.value.insert` -> `model.sharedModel.updateSource`

jtpio commented:
Thanks Carlos!

Looks like there is a small conflict to fix.
 
Also we could document this change in this section of the docs: https://jupyterlab.readthedocs.io/en/latest/extension/extension_migration.html#api-breaking-changes

jtpio commented on packages/cells/src/model.ts:
```diff
@@ -641,18 +664,19 @@ export class CodeCellModel extends CellModel implements ICodeCellModel {
       }
     }
-    this.value.changed.connect(this._onValueChanged, this);
+    // Deprecated: Tracked on this._onSharedModelChanged
+    //this.value.changed.connect(this._onValueChanged, this);
```
Leftover comment?

```suggestion
```

hbcarlos replied:
Yep! Thanks, @jtpio!

hbcarlos commented:
I'm on it!

hbcarlos commented:
I just rebased to resolve the conflicts.

fcollonval commented on packages/codeeditor/src/editor.ts:
```diff
@@ -208,16 +204,13 @@ export namespace CodeEditor {
-      sharedModel: models.ISharedText,
-      reinitialize: boolean
-    ): void;
+    switchSharedModel(sharedModel: ISharedText, reinitialize: boolean): void;
```
This will in particular avoid rebinding to the editor

hbcarlos replied:
I'm not sure yet if that's possible, at least for now, until the move feature is implemented in Yjs.
I think the main problem would be here:
https://github.com/jupyterlab/jupyterlab/blob/9ad40ceed4f74470f2ea1d99f12c883f8d460a16/packages/notebook/src/celllist.ts#L81

Still, I need to check how to handle the deduplication of the cells' list.

fcollonval reviewed (changes requested):
Thanks @hbcarlos 

I like the intention of removing duplication. But we definitely needs to rethink properly the public API.

The code is hard to understand and induce duplication that this PR addresses partly mainly because we have two data sources: `ModelDB` and `ISharedCell`. This is one side of the problem. The other one is the public API; a cell model should represent a cell model... What I mean, is that the public API expect to be the data structure:
- cell source
- cell metadata
- cell type
- specific attributes (like outputs, trusted for code cells)

The primary reason `ModelDB` was not kept hidden is RTC to get access to the collaborators list from it: https://github.com/jupyterlab/jupyterlab/pull/2058

> It would have probably be better to expose only the collaborators rather than the model. But this is off topic.

Back to the point, we should try hard to not expose the data sync or server connection directly but to provide public API that reflect the data structure and is easy to use.
For that reason, I would suggest still providing a getter/setter to a cell source attribute that is using the shared model behind the scene. Something like `ICellModel.source` would make sense to reflect the nbformat and highlight the API change. And you could probably create an helper `updateSource` to forward to the corresponding shared model.
Regarding the signal we should start using more precise signal payload in particular the `contentChanged` signal could be modified to tell which part of the model changed (like `source` or `metadata`).

hbcarlos commented:
> For that reason, I would suggest still providing a getter/setter to a cell source attribute that is using the shared model behind the scene. Something like ICellModel.source would make sense to reflect the nbformat and highlight the API change. And you could probably create an helper updateSource to forward to the corresponding shared model.
Regarding the signal we should start using more precise signal payload in particular the contentChanged signal could be modified to tell which part of the model changed (like source or metadata).

Thanks for the feedback @fcollonval. I agree the sharedModel/ModelDB should be private and the Model/DocumentModel/CellModel should provide the necessary methods to access the information on sharedModel/ModelDB.

hbcarlos commented:
The last commit implements the changes commented in https://github.com/jupyterlab/jupyterlab/pull/11601#pullrequestreview-839236143

I added a new signal, `sourceChanged` to the `CodeEditor.IModel`, triggered when the source (content) changes.
There is an important gap between the signal's arguments triggered by the sharedModel and modelDB (`@jupyterlab/observables`). Should we get closer to the observable signals (less breaking changes, but we need to change the sharedModel signal) or to the sharedModel?

Now when the source changes, we are emitting two signals content/source changed. Should we remove the generic signal `contentChanged` in favor of `sourceChanged`, `metadataChanged`, and `stateChanged`?

fcollonval commented:
Thanks @hbcarlos for the changes.

> Should we get closer to the observable signals (less breaking changes, but we need to change the sharedModel signal) or to the sharedModel?
> Now when the source changes, we are emitting two signals content/source changed. Should we remove the generic signal contentChanged in favor of sourceChanged, metadataChanged, and stateChanged?

It would be good to bring those questions to the JupyterLab Team meeting to highlight them.

hbcarlos commented:
I added some more commits.

**I made modelDB private:**
I commented (will remove later) some logic implemented in old attempts of RTC. This code is not being used and was accessing the modelDB.

**I made sharedModel private:**
There are some places where we need to access specific YJS objects:
* The code mirror editor needs a Y.Text, the Awareness, and UndoManager objects.
* The document provider needs a Y.Doc.
I added two new properties (`ymodel` in IModel and `ydocument` in DocumentModel) that expose only the strictly necessary from YJS.

**Regarding the sourceChanged signal:**
I decided to have as an argument a similar type to the `IObservableString.IChangedArgs`, making it as easier as possible to migrate for extension authors.
